### DISCUSSION TITLE
Input mysteriously stopped working

### DISCUSSION THREAD

**Discussion by ghost** [saulpw/visidata#1251] [Q&A]
I'm asking this as a discussion question, as I don't think I can reproduce this bug yet.

I'm working in VisiData, and suddenly I'm not able to input things.
(I undid and redid a few times, but this might be unrelated. I also enabled the sheet guard using the menu, which I haven't done before.)

For example:

- `^` would normally let me change the column name, but now it just erases the column name and doesn't let my type anything.
- `e` would normally let me edit a cell, but now it deletes all the cells in the column with the same value and doesn't let me edit.
- `=` would normally let me create a column from a Python expression, but now it creates a new unnamed blank column.
- `Ctrl-s` would normally let me set a filename, but now it skips immediately to "already exists. overwite? no". (But it does let me change this to "yes".
- `Space` would normally let me enter a longname command, but now it does nothing.
- Using the menu to select `System -> Python -> exec()` also doesn't work, as that would require me to input something.

This appears to have infected all my open sheets.
I've checked to make sure I don't have Num Lock or Caps Lock on!

Is there anything I can do to inspect the process and figure out what's gone wrong, so I can create a coherent bug report?

**Comment by geekscrapy** [comment]
Do you have any peripherals which do keyboard input? Do you have any of the keyboard locks on? I sometimes get caught out having Capslock on and vd consumes the UPPER case varient of the key combo which confused me a lot at first! Maybe one of your peripherals is doing the same?

**Reply by ghost** [reply]
That was my first thought. But I opened up a new terminal window, and VisiData is working perfectly in the new session. But it's still not working in the original session!

**Reply by saulpw** [reply]
How many times has this happened?

**Reply by ghost** [reply]
I've never had this happen before.

I even suspended VisiData with `Ctrl-z` and launched a new instance of VisiData in the same terminal, in case it was a bug in my terminal, but the new instance had no issues. When I resumed the suspended instance, it was still broken.

**Reply by ghost** [reply]
I've just noticed now something in the status bar that I don't know the meaning of:

![image](https://user-images.githubusercontent.com/55477181/147183244-d65b9479-dfaa-4ca3-963e-26ba89312897.png)

What does `[M]` mean?

**Reply by saulpw** [reply]
"Modified".  I don't think it's that, but I'm not sure what it could be.  Maybe a stray cosmic ray.  Hopefully the data isn't precious, but you might just have to quit and start over.

**Reply by ghost** [reply]
I'm manually retracing my steps following the command-log. Sadly I can't even save the command-log to attempt to replay it, as it automatically sets the filename to an empty string. But at least I can still view it! I'm wondering if I accidentally hit a "cursed" combination of keys. But I guess cosmic rays are a real possibility. 🌞

**Reply by anjakefala** [reply]
Can you screenshot the cmdlog?

Also, if you could look at the **Options Sheet** (`Shift+O`), and let me know if any of the options are different from default!

This is so cool!

**Reply by anjakefala** [reply]
`editText` is the function that all of the above commands has in common, if I look at that code

```
 if v is None:
         try:
             v = vd.editline(vd.activeSheet._scr, y, x, w, display=display, **kwargs)                                                             
         except AcceptInput as e:
             v = e.args[0]
```

If `AcceptInput` triggers, then it just returns the default. And one guess is that it is triggering for some reason in that session.

**Reply by ghost** [reply]
Oooh. This is interesting.

I wasn't sure if you really wanted 8 full screens of cmdlog in screenshot form, so I scrolled down to see if I could find the transition point where it got interesting.

Here's the interesting bit:

![image](https://user-images.githubusercontent.com/55477181/147188651-0012f213-5d9e-42d4-bc8c-1775c81114a6.png)

Specifically, look at the `comment` field, and compare it to the `longname` and `keystrokes` fields.

**Reply by ghost** [reply]
Non-default options are:

| option                      |  global_value                                  |
| ---------------------| ------------------------------------|
| clipboard_copy_cmd | /mnt/c/Windows/System32/clip.exe |

My `.visidatarc` is:

``` python
# Keybindings
# See https://www.visidata.org/docs/customize/
bindkey('zx', 'cut-cell')
bindkey('gzx', 'cut-cells')
bindkey('x', 'cut-row')
bindkey('gx', 'cut-selected')

# Clipboard command
options.clipboard_copy_cmd = '/mnt/c/Windows/System32/clip.exe'


# Don't clean Excel sheet names by default
options.clean_names = False


# Progress indicator
@Sheet.property
def pctRows(sheet):
    return (sheet.cursorRowIndex + 1)/(sheet.nRows or 1)

Sheet.class_options.disp_rstatus_fmt += '  {sheet.pctRows:>4.0%}'


# Sub-sheets from Frequency Sheet
# https://github.com/saulpw/visidata/discussions/1151#discussioncomment-1467379
from visidata.freqtbl import valueNames
@FreqTableSheet.api
def freq_subsheets(sheet, rows):
    for r in rows:
        vs = copy(sheet.source)
        vs.name += "_" + valueNames(r.discrete_keys, r.numeric_key)
        vs.rows = copy(r.sourcerows)
        yield vs

FreqTableSheet.addCommand('z'+ENTER, 'open-rows', 'for vs in freq_subsheets(selectedRows): vd.push(vs)',
    '(.visidatarc command) open a separate copy of source sheet with rows that are grouped in each selected frequency bin')


# Custom functions
def IF(case, a, b):
    """If case evaluates to True, return a, else return b"""
    if case:
        return a
    else:
        return b


def FREQUENCY(items):
    """Return collections.Counter of frequencies based on list of items"""
    from collections import Counter
    return Counter(items)
```

**Comment by anjakefala** [comment]
What are the chances that you are in a paused replay @davidwales?!?!? Sorry, I am so excited, this weird comment was the breakthrough, you are *totally* right.

```
 comment = vd.currentReplayRow.comment if vd.currentReplayRow else cmd.helpstr 
 ```

**Reply by anjakefala** [reply]
Can you try `Ctrl+U`? That un-pauses replays!

**Reply by anjakefala** [reply]
```
 @VisiData.api                                                                                                                                                                           
 def replayOne(vd, r):                                                                                                                                                                   
         'Replay the command in one given row.'                                                                                                                                          
         vd.currentReplayRow = r                                                                                                                                                         
         longname = getattr(r, 'longname', None)                                                                                                                                         
                                                                                                                                                                                         
         if r.sheet and longname not in ['set-option', 'unset-option']:                                                                                                                  
             vs = vd.getSheet(r.sheet) or vd.error('no sheet named %s' % r.sheet)                                                                                                        
         else:                                                                                                                                                                           
             vs = None                                                                                                                                                                                                                                                                                                                                                            
         if longname in ['set-option', 'unset-option']:                                                                                                                                  
             try:                                                                                                                                                                        
                 context = vs if r.sheet and vs else vd                                                                                                                                  
                 option_scope = r.sheet or r.col or 'global'                                                                                                                             
                 if option_scope == 'override': option_scope = 'global' # override is deprecated, is now global                                                                          
                 if longname == 'set-option':                                                                                                                                            
                     context.options.set(r.row, r.input, option_scope)                                                                                                                   
                 else:                                                                                                                                                                   
                     context.options.unset(r.row, option_scope)                                                                                                                          
                                                                                                                                                                                         
                 escaped = False                                                                                                                                                         
             except Exception as e:                                                                                                                                                      
                 vd.exceptionCaught(e)                                                                                                                                                   
                 escaped = True                                                                                                                                                          
         else:                                                                                                                                                                           
             if vs:                                                                                                                                                                      
                 vd.push(vs)                                                                                                                                                             
             else:                                                                                                                                                                       
                 vs = vd.activeSheet or vd.cmdlog                                                                                                                                        
                                                                                                                                                                                         
             vd.moveToReplayContext(r, vs)                                                                                                                                               
                                                                                                                                                                                         
             if r.comment:                                                                                                                                                               
                 vd.status(r.comment)                                                                                                                                                    
                                                                                                                                                                                         
             vd.keystrokes = r.keystrokes                                                                                                                                                
             # <=v1.2 used keystrokes in longname column; getCommand fetches both                                                                                                        
             escaped = vs.execCommand(longname if longname else r.keystrokes, keystrokes=r.keystrokes)                                                                                   
                                                                                                                                                                                         
         vd.currentReplayRow = None                                                                                                                                                      
                                                                                                                                                                                         
         if escaped:  # escape during replay aborts replay                                                                                                                               
             vd.warning('replay aborted during %s' % (longname or r.keystrokes))                                                                                                         
         return escaped  
 ```
         
      Hmmm  I am not sure if `vd.currentReplayRow` resets to `None`, if there was an exception in the replay?

**Reply by ghost** [reply]
I tried it, (and also `Ctrl-N` and `Ctrl-K` for good measure!) and it didn't change anything. But thanks for the enthusiasm. 😀

**Comment by anjakefala** [comment]
DAVID.

IF I COMMENT OUT

`vd.currentReplayRow = None` in that code block, we get your bug!! It reproduces it!

I have no idea why your `vd.currentReplayRow` is not being set to `None`!!!

**Reply by ghost** [reply]
🤯

**Reply by ghost** [reply]
When you have the bug, can you think of any way of pulling more information out about what's going on?
Or do we assume a cosmic ray has deleted that part of code from RAM, and knock off for Christmas? 🎉

**Comment by ghost** [comment]
I'm logging off for Christmas, so I won't be responding here for awhile.
Here's a last info-dump for your curiosity:

I opened up the list of all errors which have occurred while running this instance of VisiData:

![image](https://user-images.githubusercontent.com/55477181/147195903-f9d69857-4c14-4af7-8774-8d030ccf442a.png)

In order of the screenshot above:

![image](https://user-images.githubusercontent.com/55477181/147195982-c04b9fad-1478-4707-836d-a2433f5ccf17.png)
I think this might be because we're trying to save files with a blank name?

![image](https://user-images.githubusercontent.com/55477181/147196116-03fd089a-6b63-4bdf-8f71-4d3d8b147713.png)
Not sure about this one.

![image](https://user-images.githubusercontent.com/55477181/147196170-db3d8db0-21c5-4702-b822-ab5613f783d6.png)
This one happens I *think* when trying to enable profiling?

![image](https://user-images.githubusercontent.com/55477181/147196298-17cedeaa-d851-458f-a89c-563b9f8c3624.png)
I'm assuming this one happened because I tried to import a Python module, but I can't type anything...

**Reply by anjakefala** [reply]
Merry Christmas! Thank you so much for this fun puzzle and for sharing so much session data!

**Reply by anjakefala** [reply]
Unfortunately, the errors seem to all be related to those commands not having an input! But, it is valuable information that an exceptionCaught related to the initiating of the bug had not run. =)

**Reply by anjakefala** [reply]
`getLastArgs` would be why if `vd.currentReplayRow` was stuck with a row object. blank inputs were being passed to commands. So, the only open question is what is a posible context where `vd.currentReplayRow` is stuck with a row object.

**Comment by anjakefala** [comment]
Okay, so only one of your default options differed, but you had at least 3 option settings in your visidatarc.

I wonder if it stopped loading at the `options.clean_names = False`, and the way for me to confirm that is to see if I can notice the rstatus in one of your screenshots.

**Comment by anjakefala** [comment]
Your `pctRows` is showing in your status, so it is not that, and your visidatarc had loaded at least until that point. Hm!

**Comment by anjakefala** [comment]
I think it is probably important that one of your plugins failed to load (`plugins.dedupe`, and I am wondering if that is possibly related...

**Reply by anjakefala** [reply]
The (None) reference in the `row` column for the `open-row` that is likely stuck in the `vd.currentReplayRow` might have something to do with it, but still cannot reproduce!

**Comment by jazzabeanie** [comment]
I'm getting this same issue every now and then. It's as if every time I do something that needs input, enter is immediately pressed before I get a chance to do anything. I've never touched the visidatarc file, or installed plugins. I have viewed the CommandLog Sheet but never knowingly replayed it. I am using visidata as pager from psql (if that matters).

**Reply by anjakefala** [reply]
@jazzabeanie, if at any point you run into this, and you are able to get me detailed information from your **CommandLog**, that would help a lot with trying to catch this! At this point, I need a repro to try to narrow in on this.

**Comment by anjakefala** [comment]
Tentative fix reported here: https://github.com/saulpw/visidata/issues/1319#issuecomment-1116857688.

Please re-open issue #1319 if it comes up again! Fix is on `develop`, and will be part of VIsiData v2.9, once v2.9 is released.
